snuba/admin/clickhouse/migration_checks.py

```diff
@@ -0,0 +1,215 @@
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from enum import Enum
+from typing import List, Sequence, Tuple
+
+from snuba.admin.migrations_policies import get_migration_group_polices
+from snuba.migrations.groups import MigrationGroup
+from snuba.migrations.policies import MigrationPolicy
+from snuba.migrations.runner import MigrationDetails, MigrationKey
+from snuba.migrations.status import Status
+
+
+class Reason(Enum):
+    NO_REASON_NEEDED = ""
+    ALREADY_RUN = "already run"
+    NOT_RUN_YET = "can't reverse if not already run"
+    NEEDS_SUBSEQUENT_MIGRATIONS = "subsequent migrations must be reversed first"
+    NEEDS_EARLIER_MIGRATIONS = "earlier migrations must run first"
+    RUN_POLICY = "group not allowed run policy"
+    REVERSE_POLICY = "group not allowed reverse policy"
+
+
+@dataclass
+class ResultReason:
+    allowed: bool
+    reason: Reason
+
+
+@dataclass
+class MigrationData:
+    migration_id: str
+    status: str
+    blocking: bool
+    can_run: bool
+    can_reverse: bool
+    run_reason: str
+    reverse_reason: str
+
+
+class Checker(ABC):
+    """
+    A checker is used to encapsulate logic about whether
+    a migration can be run or reversed based on some criteria
+    or state the checker has.
+
+    If a migration can be run/reversed, the ResultReason that
+    is returned should have allowed = True and
+    reason = Reason.NO_REASON_NEEDED.
```

Comment:
Why both? `allowed` should be enough. In order to enforce that, please add a post_init method to the `ResultReason` class so that you cannot provide a reason if `allowed == True` and you must provide a reason if `allowed==False`